### ISSUE TITLE
Data file uploads failing

### ISSUE THREAD

**Issue by hackdna** [parklab/refinery-platform#1282]
Commit: c6aae7951b62e70fded561d76e1c65c9b2f4c281
Browser: Chrome 51.0.2704.103 or the latest 52.0.2743.82

### Steps to reproduce
Upload a file using [Data file upload form](https://beta.stemcellcommons.org/data_set_manager/import/)

### Observed behavior
File is not uploaded. Browser console shows:
```
Uncaught SyntaxError: Failed to execute 'importScripts' on 'WorkerGlobalScope': The URL '/static/vendor/spark-md5/spark-md5.min.js' is invalid.onmessage @ blob:https://beta.stemcellcommons.org/6bc257f1-d06a-4037-a1aa-f930749283db:1
```
### Expected behavior
File uploaded, no errors in the console.

**Comment by scottx611x** [comment]
Chrome `51.0.2704.103`:
```
Uncaught SyntaxError: Failed to execute 'importScripts' on 'WorkerGlobalScope': The URL '/static/vendor/spark-md5/spark-md5.min.js' is invalid.
```

Firefox `47.0.1`:
`SyntaxError: Failed to load worker script at "/static/vendor/spark-md5/spark-md5.min.js"`

**Comment by ngehlenborg** [comment]
Tested on Mac OS X 10.11.5 in Safari Version 9.1.1 (11601.6.17) and failed with 

```JavaScript
[Error] ReferenceError: Can't find variable: File (blob:https://test.stemcellcommons.org/8f3ae200-3b08-49c3-b76b-ce1d2c75bcb7, line 1)
```

**Comment by hackdna** [comment]
Firefox 39.0:
```
10:20:25.990 : SyntaxError: An invalid or illegal string was specified1 316d20b4-1d76-8c40-91bf-6e21e4b02ce2:1:0
```

**Comment by scottx611x** [comment]
`test.stemcellcommons.org` is working fine for both my Firefox and Chrome!

**Comment by hackdna** [comment]
Firefox 42.0:
```
10:23:31.160 : InvalidStateError: An attempt was made to use an object that is not, or is no longer, usable1 2714513d-124a-bb44-9b68-f43cb544ada1:1:0
```

**Comment by ngehlenborg** [comment]
@scottx611x: I am testing on http://test.stemcellcommons.org and it is not working for me.

**Comment by ngehlenborg** [comment]
Tested http://test.stemcellcommons.org (29745d057ba2ee5e4da65c05e74f3ee31915be1d) on Mac OS X 10.11.5 in Chrome Version 52.0.2743.82 (64-bit) and succeeded.

**Comment by ngehlenborg** [comment]
Tested http://beta.stemcellcommons.org (c6aae7951b62e70fded561d76e1c65c9b2f4c281) on Mac OS X 10.11.5 in Chrome Version 52.0.2743.82 (64-bit) and failing with 

```JavaScript
Uncaught SyntaxError: Failed to execute 'importScripts' on 'WorkerGlobalScope': The URL '/static/vendor/spark-md5/spark-md5.min.js' is invalid.onmessage @ blob:https://beta.stemcellcommons.org/7200c2c2-7452-4d81-8df0-c6ccdc056029:1
```

**Comment by hackdna** [comment]
Safari 9.1.2:
```
SyntaxError: DOM Exception 12: An invalid or illegal string was specified.
```

**Comment by hackdna** [comment]
Upload failed on test.stemcellcommons.org in Firefox 45 and Chrome 49.

**Comment by scottx611x** [comment]
<a href="https:&#x2F;&#x2F;trello.com&#x2F;c&#x2F;ZRFFO1As&#x2F;67-test"><img src="https:&#x2F;&#x2F;github.trello.services&#x2F;images&#x2F;trello-icon.png" width="12" height="12"> test</a>

**Comment by scottx611x** [comment]
<a href="https:&#x2F;&#x2F;trello.com&#x2F;c&#x2F;XQkfOVOB&#x2F;293-test"><img src="https:&#x2F;&#x2F;github.trello.services&#x2F;images&#x2F;trello-icon.png" width="12" height="12"> test</a>

**Comment by ngehlenborg** [comment]
Ok, it looks like this might be due to an issue with File API support in Safari (as the latest Chrome and  Firefox appear to perform fine, at least on Mac OS X):

1. On http://test.stemcellcommons.org I can get it to work with Firefox 47 and Chrome 52 but not with Safari 9.1.2.
1. It seems that the File API is not fully supported by Safari 9.1.x, hence the error message “[Error] ReferenceError: Can't find variable: File (blob:https://test.stemcellcommons.org/ec6dfb2c-3052-4ae1-879c-061ea56b7905, line 1)”. The part that is missing is support for the File constructor.
1. According to caniuse.com (http://caniuse.com/#feat=fileapi), Firefox added support for the File API only in version 47, Safari will have full support in version 10.

@hackdna: can you double check that it is indeed Chrome 49 that is failing on Linux and if so, please provide the full version number? During March 2016 there were 5 stable releases of Chrome 49 (for Linux, Windows, and Mac) according to http://googlechromereleases.blogspot.com/search/label/Stable%20updates?updated-max=2016-03-24T10:55:00-07:00&max-results=20&start=22&by-date=false.

**Comment by ngehlenborg** [comment]
@flekschas: The issue in Safari is triggered by 

```JavaScript      
var slice = File.prototype.slice || File.prototype.mozSlice || File.prototype.webkitSlice;
```

Here:  https://github.com/parklab/refinery-platform/blob/2982a0ebd78db01bbb679647622fd601dda27173/refinery/ui/source/js/data-set-import/controllers/file-upload.js#L58

and possibly also here:

https://github.com/parklab/refinery-platform/blob/2982a0ebd78db01bbb679647622fd601dda27173/refinery/ui/source/js/data-set-import/controllers/file-upload.js#L120

Is there any other way to call the `slice` method on the `file` object?

```JavaScript
reader.readAsArrayBuffer(slice.call(file, startIndex, end));
```

**Comment by ngehlenborg** [comment]
Looks like this is even more complicated: I can run the SparkMD5 demo in my Safari 9.1.2 although it contains the exact same line of code as the one that fails in Refinery:

http://9px.ir/demo/incremental-md5.html

The only difference is that Refinery runs the code in a WebWorker, while the demo does not.

**Comment by ngehlenborg** [comment]
And there is a 4 year old Chromium bug report that illustrates the issue (still reproducible with Safari 9.1.2):

https://bugs.chromium.org/p/chromium/issues/detail?id=147503

**Comment by hackdna** [comment]
Results of data file upload testing with different OS/browser combinations: https://docs.google.com/spreadsheets/d/1ZMabz-jgra8wzC43fU__J7wn77T1eE2v0FSh1ngYRfo

**Comment by ngehlenborg** [comment]
There is a workaround to get this to work in Safari 9.1.2:

http://stackoverflow.com/a/13697119/1371194

When we replace the `File.prototype.slice` etc. in https://github.com/parklab/refinery-platform/blob/2982a0ebd78db01bbb679647622fd601dda27173/refinery/ui/source/js/data-set-import/controllers/file-upload.js#L58 (but not in L120) with `Blob.prototype.slice`, the checksum is computed correctly and the file can be uploaded.

**Comment by hackdna** [comment]
<a href="https:&#x2F;&#x2F;trello.com&#x2F;c&#x2F;nppQOdjX&#x2F;187-update-data-import-ui"><img src="https:&#x2F;&#x2F;github.trello.services&#x2F;images&#x2F;trello-icon.png" width="12" height="12"> Update data import UI</a>

**Comment by flekschas** [comment]
@nils implemented and tested your suggested fix using Blob. This seems to be a strange bug in Safari related to the File API because somehow the whole File API seems to be missing in webworkers while Safari itself supports the API since version 6.

**Comment by ngehlenborg** [comment]
@flekschas Thanks! @jkmarx and I also tried it and it seems to work. I filed a bug report with Apple regarding this bug.

**Comment by flekschas** [comment]
By the way Firefox partially supports the File API since version 3.6 and fully since version 28 and Safari is generally only lacking the File constructor (which we don't use as far as I know).

**Comment by jkmarx** [comment]
Waiting for @hackdna to confirm fix.
Resolved by commit #62824e0a39617bd3bfb279e8b17b51a34be5a653
https://github.com/parklab/refinery-platform/commit/62824e0a39617bd3bfb279e8b17b51a34be5a653

**Comment by hackdna** [comment]
This is resolved but #1292 remains.
